File: skiko/skiko-ganesh/src/nativeJsMain/cpp/ImageGanesh.cc
#include "SkImage.h"
#include "SkBitmap.h"
#include "ganesh/GrBackendSurface.h"
#include "ganesh/GrDirectContext.h"
#include "include/gpu/ganesh/SkImageGanesh.h"
#include "common.h"

SKIKO_EXPORT KBoolean org_jetbrains_skia_Image__1nReadPixelsBitmap
  (KNativePointer ptr, KNativePointer contextPtr, KNativePointer bitmapPtr, KInt srcX, KInt srcY, KBoolean cache) {
    SkImage* instance = reinterpret_cast<SkImage*>(ptr);
    GrDirectContext* context = reinterpret_cast<GrDirectContext*>(contextPtr);
    SkBitmap* bitmap = reinterpret_cast<SkBitmap*>(bitmapPtr);
    auto cachingHint = cache ? SkImage::CachingHint::kAllow_CachingHint : SkImage::CachingHint::kDisallow_CachingHint;
    return instance->readPixels(context, bitmap->info(), bitmap->getPixels(), bitmap->pixmap().rowBytes(), srcX, srcY, cachingHint);
}

SKIKO_EXPORT KNativePointer org_jetbrains_skia_Image__1nAdoptTextureFrom
  (KNativePointer contextPtr, KNativePointer backendTexturePtr, KInt surfaceOrigin, KInt colorType) {
    GrDirectContext* context = reinterpret_cast<GrDirectContext*>(contextPtr);
    GrBackendTexture* backendTexture = reinterpret_cast<GrBackendTexture*>(backendTexturePtr);

    sk_sp<SkImage> image = SkImages::AdoptTextureFrom(
        static_cast<GrRecordingContext*>(context),
        *backendTexture,
        static_cast<GrSurfaceOrigin>(surfaceOrigin),
        static_cast<SkColorType>(colorType)
    );

    return reinterpret_cast<KNativePointer>(image.release());
}

File: skiko/skiko-ganesh/src/commonMain/kotlin/org/jetbrains/skia/ImageGanesh.kt
package org.jetbrains.skia

import org.jetbrains.skia.impl.*

fun Image.Companion.adoptTextureFrom(
    context: DirectContext,
    backendTexture: BackendTexture,
    origin: SurfaceOrigin,
    colorType: ColorType,
): Image {
    return try {
        Stats.onNativeCall()
        val ptr = _nAdoptTextureFrom(
            getPtr(context),
            getPtr(backendTexture),
            origin.ordinal,
            colorType.ordinal
        )
        if (ptr == Native.NullPointer) throw RuntimeException("Failed to Image::makeFromTexture")
        Image(ptr)
    } finally {
        reachabilityBarrier(context)
        reachabilityBarrier(backendTexture)
    }
}

fun Image.readPixels(context: DirectContext, dst: Bitmap): Boolean =
    readPixels(context, dst, 0, 0, false)

fun Image.readPixels(context: DirectContext, dst: Bitmap, srcX: Int, srcY: Int): Boolean =
    readPixels(context, dst, srcX, srcY, false)

fun Image.readPixels(context: DirectContext?, dst: Bitmap, srcX: Int, srcY: Int, cache: Boolean): Boolean =
    readPixelsWithContextPtr(getPtr(context), dst, srcX, srcY, cache)
